tests/runner/test_runner.py

```diff
@@ -344,12 +387,102 @@ async def test_runner_can_cancel_flow_runs(
         assert "This flow was cancelled!" in caplog.text
 
     @pytest.mark.usefixtures("use_hosted_api_server")
-    async def test_runner_runs_on_crashed_hooks(
+    @pytest.mark.flaky
+    async def test_runner_runs_on_cancellation_hooks_for_remotely_stored_flows(
+        self, prefect_client: PrefectClient, caplog
+    ):
+        runner = Runner(query_seconds=2)
+
+        storage = MockStorage()
+        storage.code = dedent(
+            """\
+            from time import sleep
+
+            from prefect import flow
+            from prefect.logging.loggers import flow_run_logger
+
+            def on_cancellation(flow, flow_run, state):
+                logger = flow_run_logger(flow_run, flow)
+                logger.info("This flow was cancelled!")
+
+            @flow(on_cancellation=[on_cancellation], log_prints=True)
+            def cancel_flow(sleep_time: int = 100):
+                sleep(sleep_time)
+            """
+        )
+
+        deployment_id = await runner.add_flow(
+            await flow.from_source(source=storage, entrypoint="flows.py:cancel_flow"),
+            name=__file__,
+        )
+
+        async with anyio.create_task_group() as tg:
+            tg.start_soon(runner.start)
+
+            flow_run = await prefect_client.create_flow_run_from_deployment(
+                deployment_id=deployment_id
+            )
+
+            # Need to wait for polling loop to pick up flow run and
+            # start execution
+            for _ in range(15):
+                await anyio.sleep(1)
+                flow_run = await prefect_client.read_flow_run(flow_run_id=flow_run.id)
+                if flow_run.state.is_running():
+                    break
+
+            await prefect_client.set_flow_run_state(
+                flow_run_id=flow_run.id,
+                state=flow_run.state.copy(
+                    update={"name": "Cancelling", "type": StateType.CANCELLING}
+                ),
+            )
+
+            # Need to wait for polling loop to pick up flow run and then
+            # finish cancellation
+            for _ in range(15):
+                await anyio.sleep(1)
+                flow_run = await prefect_client.read_flow_run(flow_run_id=flow_run.id)
+                if flow_run.state.is_cancelled():
+                    break
+
+            await runner.stop()
+            tg.cancel_scope.cancel()
+
+        assert flow_run.state.is_cancelled()
+        # check to make sure on_cancellation hook was called
+        assert "This flow was cancelled!" in caplog.text
```

Comment:
Should we assert that the `on_cancellation` hook was called here?

Reply:
The assertion for `caplog.text` matches the message the state change hook will log when called. I couldn't use mocks because I'd have to declare the mock at the top level of the module. That's also where the flows and state change hooks need to be declared to be loadable by the runner. It's not ideal, but these tests do fail if the state change hook isn't called.

Reply:
sounds good, thanks for clarifying 👍